File: app/src/main/java/com/ed/redditapp/ui/subreddit/SubredditPostsAsyncTask.java
package com.ed.redditapp.ui.subreddit;

import com.ed.redditapp.lib.AsyncTask;
import com.ed.redditapp.lib.api.RedditApi;
import com.ed.redditapp.ui.postlist.Post;

public class SubredditPostsAsyncTask {
    private final RedditApi redditApi;

    public interface Callback {
        void callback(Post[] posts);
    }

    public SubredditPostsAsyncTask(RedditApi redditApi) {
        this.redditApi = redditApi;
    }

    public void run(String subredditName, Callback callback) {
        new AsyncTask().runAsync(h -> {
            Post[] posts = redditApi.getSubredditPosts(subredditName);
            h.onUIThread(() -> callback.callback(posts));
        });
    }
}

File: app/src/main/java/com/ed/redditapp/ui/subreddit/SubredditPostsAsyncTask.kt
package com.ed.redditapp.ui.subreddit

import com.ed.redditapp.lib.AsyncTask
import com.ed.redditapp.lib.api.Post
import com.ed.redditapp.lib.api.RedditApi

typealias SubredditPostsCallback = (posts: Array<Post>) -> Unit

class SubredditPostsAsyncTask(private val redditApi: RedditApi) {
    fun run(subredditName: String, callback: SubredditPostsCallback) {
        AsyncTask().runAsync { h ->
            val posts: Array<Post> = redditApi.getSubredditPosts(subredditName)
            h.onUIThread { callback(posts) }
        }
    }
}